src/System.Runtime.Analyzers/Core/ProvideCorrectArgumentsToFormattingMethods.cs

```diff
@@ -4,36 +4,421 @@
+
+                        return;
+                    }
+
+                    // params case
+                    var paramsArgument = invocation.ArgumentsInParameterOrder[info.FormatStringIndex + 1];
+                    if (paramsArgument.Kind != ArgumentKind.ParamArray)
+                    {
+                        // wrong format
+                        return;
+                    }
+
+                    var arrayCreation = paramsArgument.Value as IArrayCreationExpression;
+                    if (arrayCreation == null ||
+                        !object.Equals(arrayCreation.ElementType, formatInfo.Object) ||
+                        arrayCreation.DimensionSizes.Length != 1)
+                    {
+                        // wrong format
+                        return;
+                    }
+
+                    // explictly giving object array case.
+                    // this should go away once the bug is fixed
+                    if (invocation.Syntax.Language == LanguageNames.CSharp &&
+                        invocation.ArgumentsInParameterOrder.Count(a => a.Kind == ArgumentKind.ParamArray) ==
+                        invocation.ArgumentsInSourceOrder.Count(a => a.Kind == ArgumentKind.ParamArray))
+                    {
+                        // TODO: due to a bug - https://github.com/dotnet/roslyn/issues/7342
+                        //       currently this (explictly giving object array case) is not supported.
+                        //
+                        //       this check is only for CSharp, since VB correctly normalize explicit
+                        //       and implicit params array to one format. but VB has different issue where one can't
+                        //       distinguish explicit and implicit params argument since it is already normalized in one
+                        //       format in both ArgumentsInParameterOrder and ArgumentsInSourceOrder.
+                        //
+                        return;
+                    }
+
+                    // compiler generating object array for params case
+                    var dimensionInitializer = arrayCreation.ElementValues as IDimensionArrayInitializer;
+                    if (dimensionInitializer == null)
+                    {
+                        // unsupported format
+                        return;
+                    }
+
+                    // REVIEW: "ElementValues" is a bit confusing where I need to double dot those to get number of elements
+                    var actualArgumentCount = dimensionInitializer.ElementValues.Length;
+                    if (actualArgumentCount != expectedStringFormatArgumentCount)
+                    {
+                        operationContext.ReportDiagnostic(operationContext.Operation.Syntax.CreateDiagnostic(Rule));
+                    }
+                }, OperationKind.InvocationExpression);
+            });
+        }
+
+        private int GetFormattingArguments(string format)
```

Comment:
Was this code picked up from the corefx repo's implementation of string.Format? If so, can you add a comment\link to that?